src/main/scala/com/scality/clueso/merge/TableFilesMerger.scala

```diff
@@ -0,0 +1,200 @@
+
+
+  case class MergeInstructions(numPartitions : Int)
+
+  def checkMergeEligibility(getPath: Path): Option[MergeInstructions] = {
+    // number of records criteria
+    val numberOfFiles = fs.listStatus(getPath, parquetFilesFilter).length
+
+    if (numberOfFiles >= config.mergeMinFiles) {
+      val data = spark.read
+          .parquet(getPath.toUri.toString)
+
+      val numRecords = data.count
+      val numFinalFiles = Math.floor(numRecords / config.mergeFactor).toInt + 1
+
+      if (numFinalFiles < numberOfFiles) {
+        return Some(MergeInstructions(numFinalFiles))
+      }
+    }
+
+    None
+  }
+
+
+  val partDirnamePattern = "([A-Za-z0-9_]+)=(.*)".r
+
+  def merge() = {
+    // go thru all partitions and see which ones are illegible for merging
+    val landingPartitionsIt = fs.listStatusIterator(new Path(config.landingPath))
+
+    while (landingPartitionsIt.hasNext) {
+      val fileStatus = landingPartitionsIt.next()
+
+      if (fileStatus.isDirectory &&
+        fileStatus.getPath.getName.matches(partDirnamePattern.regex)) {
+
+        checkMergeEligibility(fileStatus.getPath) match {
+          case Some(MergeInstructions(numPartitions)) => {
+            val regexMatch = partDirnamePattern.findFirstMatchIn(fileStatus.getPath.getName)
+
+            regexMatch.map { rm =>
+              mergePartition(partitionColumnName = rm.group(1),
+                partitionValue = rm.group(2),
+                numPartitions)
+            }
+          }
+          case _ =>
+        }
+
+
+      }
+    }
+  }
+
+  def writeMergedData(mergedData: Dataset[Row], outputPath: String) = {
+    mergedData.write
+      .parquet(outputPath)
+  }
+
+  def replaceStagingWithMerged(stagingPartitionPath: String, outputPath: String) = {
+    // list all files in staging
+    val stagingParquetFiles = fs.listStatus(new Path(stagingPartitionPath), parquetFilesFilter)
+
+    // move all from merge path to staging
+    val mergedFiles = fs.listStatus(new Path(outputPath), parquetFilesFilter)
+
+    mergedFiles.foreach(mf => fs.rename(mf.getPath, new Path(s"$stagingPartitionPath/${mf.getPath.getName}")))
+
+    // delete files that were on staging
+    stagingParquetFiles.foreach(sf => fs.delete(sf.getPath, false))
+  }
+
+  def removeProcessedFromLanding(landingPartitionPath: String, startTs: Long) = {
+    val landedParquetFiles = fs.listStatus(new Path(landingPartitionPath), parquetFilesFilter)
+    val mergedLandingFiles = landedParquetFiles.filter(lf => lf.getModificationTime < startTs - config.landingPurgeTolerance.toMillis)
+
+    mergedLandingFiles.foreach(lf => fs.delete(lf.getPath, false))
+  }
+
+  def mergePartition(partitionColumnName : String, partitionValue : String, numPartitions : Int) = {
+    val mergeOutputId = UUID.randomUUID()
+    val mergePath = s"${config.mergePath}/$mergeOutputId"
+    val outputPath = s"$mergePath/merged_data"
+
+    val lockFilePath = lockPath(mergePath)
+
+    if (acquireLock(lockFilePath)) {
```

Comment:
this is true if the operation to create the path succeeds? so, it's really just a "lock" based on a random path? Note that if you are overwriting an existing key, that will not result in an error.